java/ql/src/experimental/CWE-532/SensitiveInfoLog.ql

```diff
@@ -0,0 +1,64 @@
+/**
+ * @id java/sensitiveinfo-in-logfile
+ * @name Insertion of sensitive information into log files
+ * @description Writing sensitive information to log files can give valuable guidance to an attacker or expose sensitive user information.
+ * @kind problem
+ * @tags security
+ *       external/cwe-532
+ */
+
+import java
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow
+import PathGraph
+
+/**
+ * Gets a regular expression for matching names of variables that indicate the value being held is a credential
+ */
+private string getACredentialRegex() {
+  result = "(?i).*pass(wd|word|code|phrase)(?!.*question).*" or
+  result = "(?i)(.*username|url).*"
+}
+
+/** The variable or concatenated string with the variable that keeps sensitive information judging by its name * */
+class CredentialExpr extends Expr {
+  CredentialExpr() {
+    exists(Variable v |
+      (this.(AddExpr).getAnOperand() = v.getAnAccess() or this = v.getAnAccess())
+    |
+      v.getName().regexpMatch(getACredentialRegex())
+    )
+  }
+}
+
+/** Class of popular logging utilities * */
+class LoggerType extends RefType {
+  LoggerType() {
+    this.hasQualifiedName("org.apache.log4j", "Category") or //Log4J
+    this.hasQualifiedName("org.slf4j", "Logger") //SLF4j and Gradle Logging
+  }
+}
+
+predicate isSensitiveLoggingSink(DataFlow::Node sink) {
+  exists(MethodAccess ma |
+    ma.getMethod().getDeclaringType() instanceof LoggerType and
+    (ma.getMethod().hasName("debug") or ma.getMethod().hasName("trace")) and
+    sink.asExpr() = ma.getAnArgument()
+  )
+}
+
+class LoggerConfiguration extends DataFlow::Configuration {
+  LoggerConfiguration() { this = "Logger Configuration" }
+
+  override predicate isSource(DataFlow::Node source) { source.asExpr() instanceof CredentialExpr }
+
+  override predicate isSink(DataFlow::Node sink) { isSensitiveLoggingSink(sink) }
+
+  override predicate isAdditionalFlowStep(DataFlow::Node node1, DataFlow::Node node2) {
+    TaintTracking::localTaintStep(node1, node2)
+  }
+}
+
+from LoggerConfiguration cfg, DataFlow::Node source, DataFlow::Node sink
+where cfg.hasFlow(source, sink)
+select "Outputting sensitive information in ", sink, "to log file"
```

Comment:
The basic format of `select` for a `problem` query is first an entity which should provide the location for the result, then a message, and then a number of entity-string pairs if the message includes placeholders.  So If you want to report the alert at the sink you can do
```
select sink, "Outputting $@ to log.", source, "sensitive information"
```
or if you want to report it at the source you can do
```
select source, "Outputting sensitive information to $@.", sink, "log"
```
Now, to make it a path-problem query, you'll need to change `@kind problem` to `@kind path-problem` in the meta-data at the top, import `PathGraph` (already done), and adjust the `select` slightly as `path-problem` queries have two more components on the `select` format.  For `path-problem` queries the source-sink pair to generate the path from should be between the location entity and the message, and these should be of the type `PathNode`.  So you might write:
```
from LoggerConfiguration cfg, DataFlow::PathNode source, DataFlow::PathNode sink
where cfg.hasFlowPath(source, sink)
select sink.getNode(), source, sink, "Outputting $@ to log.", source.getNode(), "sensitive information"
```

Reply:
Hi @aschackmull ,

Thanks a lot for the detailed instructions. I've converted the query to a path-problem query and committed the code.

@luchua-bc

Reply:
Hi @aschackmull ,

Thanks a lot for the detailed instructions. I've converted the query to a path-problem query and committed the code.

@luchua-bc